File: concurrent_scraper_py/scraper/processor.py
from bs4 import BeautifulSoup
from concurrent.futures import ThreadPoolExecutor
import threading
from typing import List, Dict, Optional, NamedTuple
import time
from utils.metrics import Metrics

class ProcessedContent(NamedTuple):
    """Container for processed content."""
    title: str
    links: List[str]
    text_content: str

class ProcessError(Exception):
    """Custom exception for processing errors."""
    pass

class Processor:
    def __init__(self):
        """Initialize processor with metrics."""
        self.metrics = Metrics()
        self._lock = threading.Lock()  # Required due to GIL limitations

    def process_contents(self, contents: List[str]) -> List[ProcessedContent]:
        """
        Process multiple HTML contents concurrently using threads.
        Note: Due to Python's GIL, this won't achieve true parallelism
        since HTML parsing is CPU-bound.
        """
        self.metrics.start_operation()
        results = []

        # Using ThreadPoolExecutor, but due to GIL, this won't provide
        # true parallelism for CPU-bound tasks
        with ThreadPoolExecutor() as executor:
            future_to_content = {
                executor.submit(self._process_content, content): i 
                for i, content in enumerate(contents)
            }

            for future in future_to_content:
                try:
                    result = future.result()
                    results.append(result)
                except Exception as e:
                    results.append(e)

        return results

    def _process_content(self, html: str) -> ProcessedContent:
        """
        Process a single HTML content.
        Note: This is CPU-bound and will be limited by the GIL.
        """
        start_time = time.time()

        try:
            # Parse HTML - This is CPU-bound and will be affected by GIL
            soup = BeautifulSoup(html, 'html.parser')

            # Extract title
            title = soup.title.string if soup.title else "No title"

            # Extract links - More CPU-bound operations
            links = [
                a.get('href') 
                for a in soup.find_all('a', href=True)
            ]

            # Extract text content - More CPU-bound operations
            text_content = ' '.join(
                p.get_text() 
                for p in soup.find_all(['p', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6'])
            )

            duration = time.time() - start_time
            
            # Need lock due to shared metrics object
            with self._lock:
                self.metrics.record_success(
                    url="processing",
                    duration=duration,
                    bytes_count=len(html.encode())
                )

            return ProcessedContent(
                title=title,
                links=links,
                text_content=text_content
            )

        except Exception as e:
            duration = time.time() - start_time
            with self._lock:
                self.metrics.record_failure(
                    url="processing",
                    duration=duration
                )
            raise ProcessError(f"Failed to process content: {str(e)}")

    def get_metrics(self) -> str:
        """Get metrics summary."""
        return self.metrics.get_summary()


# For testing
if __name__ == "__main__":
    test_html = """
        <html>
            <head><title>Test Page</title></head>
            <body>
                <h1>Hello World</h1>
                <p>Test content</p>
                <a href="https://example.com">Link</a>
            </body>
        </html>
    """
    
    processor = Processor()
    results = processor.process_contents([test_html])
    
    print("\nResults:")
    for i, result in enumerate(results):
        if isinstance(result, Exception):
            print(f"Content {i+1}: Error - {result}")
        else:
            print(f"Content {i+1}:")
            print(f"  Title: {result.title}")
            print(f"  Links: {len(result.links)}")
            print(f"  Content length: {len(result.text_content)} chars")
    
    print("\nMetrics:")
    print(processor.get_metrics())

File: concurrent_scraper_rs/src/scraper/processor.rs
use rayon::prelude::*;
use scraper::{Html, Selector};
use std::sync::Arc;
use parking_lot::RwLock;
use thiserror::Error;
use crate::utils::metrics::Metrics;

#[derive(Error, Debug)]
pub enum ProcessError {
    #[error("Failed to parse HTML")]
    ParseError,
    #[error("Failed to find content")]
    ContentNotFound,
}

pub struct Processor {
    metrics: Arc<RwLock<Metrics>>,
}

#[derive(Debug)]
pub struct ProcessedContent {
    pub title: String,
    pub links: Vec<String>,
    pub text_content: String,
}

impl Processor {
    pub fn new() -> Self {
        Self {
            metrics: Arc::new(RwLock::new(Metrics::new())),
        }
    }

    pub fn process_contents(&self, contents: Vec<String>) -> Vec<Result<ProcessedContent, ProcessError>> {
        self.metrics.write().start_operation();

        contents.par_iter() // Using Rayon's parallel iterator
            .map(|content| {
                let start = std::time::Instant::now();
                
                match self.extract_content(content) {
                    Ok(processed) => {
                        self.metrics.write().record_success(
                            "processing".to_string(),
                            start.elapsed(),
                            content.len()
                        );
                        Ok(processed)
                    }
                    Err(e) => {
                        self.metrics.write().record_failure(
                            "processing".to_string(),
                            start.elapsed()
                        );
                        Err(e)
                    }
                }
            })
            .collect()
    }

    fn extract_content(&self, html: &str) -> Result<ProcessedContent, ProcessError> {
        let document = Html::parse_document(html);

        // Extract title
        let title_selector = Selector::parse("title").map_err(|_| ProcessError::ParseError)?;
        let title = document
            .select(&title_selector)
            .next()
            .ok_or(ProcessError::ContentNotFound)?
            .text()
            .collect::<String>();

        // Extract links
        let link_selector = Selector::parse("a[href]").map_err(|_| ProcessError::ParseError)?;
        let links: Vec<String> = document
            .select(&link_selector)
            .filter_map(|element| element.value().attr("href"))
            .map(String::from)
            .collect();

        // Extract text content
        let text_selector = Selector::parse("p, h1, h2, h3, h4, h5, h6")
            .map_err(|_| ProcessError::ParseError)?;
        let text_content: String = document
            .select(&text_selector)
            .map(|element| element.text().collect::<String>())
            .collect::<Vec<String>>()
            .join("\n");

        Ok(ProcessedContent {
            title,
            links,
            text_content,
        })
    }

    pub fn get_metrics(&self) -> String {
        self.metrics.read().get_summary()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_content_processing() {
        let processor = Processor::new();
        let html_content = r#"
            <html>
                <head><title>Test Page</title></head>
                <body>
                    <h1>Hello World</h1>
                    <p>Test content</p>
                    <a href="https://example.com">Link</a>
                </body>
            </html>
        "#.to_string();

        let results = processor.process_contents(vec![html_content]);
        assert_eq!(results.len(), 1);
        
        if let Ok(content) = &results[0] {
            assert_eq!(content.title, "Test Page");
            assert_eq!(content.links.len(), 1);
            assert!(content.text_content.contains("Hello World"));
        }

        println!("{}", processor.get_metrics());
    }
}
